### PR TITLE
Users can capture conditionals using `qml.cond`

### PR DESCRIPTION
**Pull request by PietropaoloFrisoni**
**Context:** The purpose of this PR is to allow `qml.cond` to be captured into plxpr, although with some restrictions on its usage (see below for details).

**Description of the Change:** Now, `qml.cond` can be captured into plxpr. The function's behavior resembles the one provided in Catalyst, with the executed branch determined at runtime.

**Benefits:** If `qml.capture.enabled()` is `True`, `qml.cond` can be used.

**Possible Drawbacks:** Here there's a list of necessary restrictions at this stage.
 
- The arguments provided to the branches must be the same for all branches.

- The arguments provided to branches must be JAX-compatible. For example, this example would not work:

```
qml.capture.enable()

dev = qml.device("default.qubit", wires=3)

@qml.qnode(dev)
def circuit(pred, wires):
    def true_fn(wires):
        qml.RY(0.1, wires=wires)
    qml.cond(pred > 0, true_fn)(wires)
    return qml.expval(qml.PauliZ(wires=0))
circuit(1, qml.wires.Wires([0]))
```

because wires are not a valid JAX type, and the arguments provided to the branch are traced.

- Tests on nested conditionals have not been included as they belong to a different (future) epic, but in principle they should work 

- If a branch returns one or more variables, every other branch must return the same abstract values.

For example, the following example would not work:

```
qml.capture.enable()

dev = qml.device("default.qubit", wires=3)

@qml.qnode(dev)
def circuit(pred, arg):
    def true_fn(arg):
        return qml.RY(arg, wires=0)
    def false_fn(arg):
        return qml.RZ(arg, wires=0), 5
    qml.cond(pred > 0, true_fn, false_fn)(arg)
    return qml.expval(qml.PauliZ(wires=0))
```

because `true_fn` and `false_fn` do not return the same abstract values. This is similar to Catalyst, where values produced inside the scope of a conditional can be returned to the outside context, but the return type signature of each branch must be identical.

- The function cannot branch on mid-circuit measurements, because these cannot be captured into plxpr yet.

**Related GitHub Issues:** None.

**Related Shortcut Stories** 
[sc-66774]
[sc-69642]

### CODE REVIEW

File: pennylane/ops/op_math/condition.py [hunk 1]
```diff
@@ -379,6 +391,10 @@ def qnode(a, x, y, z):
 
         return cond_func
 
+    if qml.capture.enabled():
+        print("Capture mode for cond")
```
Comment (PietropaoloFrisoni) [inline pennylane/ops/op_math/condition.py:395]: Sorry about this, just a debug message that I'll remove

File: pennylane/ops/op_math/condition.py [hunk 2]
```diff
@@ -430,3 +446,128 @@ def wrapper(*args, **kwargs):
...
+        def false_branch(args):
+            if jaxpr_false is not None:
+                return run_jaxpr(jaxpr_false, *args)
+            return ()
+
+        if condition:
+            return true_branch(args)
+        if elifs_conditions.size > 0:
+            return elif_branch(args, elifs_conditions, jaxpr_elifs)
+        return false_branch(args)
```
Comment (dime10) [inline pennylane/ops/op_math/condition.py:490]: More from a code design point, but the fact `elif_branch` can also call the `false_branch` is a bit unexpected.

File: pennylane/ops/op_math/condition.py [hunk 3]
```diff
@@ -430,3 +446,128 @@ def wrapper(*args, **kwargs):
...
+    elifs = (elifs,) if len(elifs) > 0 and not isinstance(elifs[0], tuple) else elifs
+
+    @wraps(true_fn)
+    def new_wrapper(*args, **kwargs):
+
+        jaxpr_true = jax.make_jaxpr(functools.partial(true_fn, **kwargs))(*args)
+        jaxpr_false = (
+            (jax.make_jaxpr(functools.partial(false_fn, **kwargs))(*args) if false_fn else None)
+            if false_fn
+            else None
```
Comment (dime10) [inline pennylane/ops/op_math/condition.py:548]: There is a double `if false_fn else None` here.

```suggestion
            jax.make_jaxpr(functools.partial(false_fn, **kwargs))(*args)
            if false_fn
            else None
```

File: pennylane/ops/op_math/condition.py [hunk 4]
```diff
@@ -430,3 +446,128 @@ def wrapper(*args, **kwargs):
...
+        elifs_conditions = []
+        jaxpr_elifs = []
+
+        for pred, elif_fn in elifs:
+            elifs_conditions.append(pred)
+            jaxpr_elifs.append(jax.make_jaxpr(functools.partial(elif_fn, **kwargs))(*args))
+
+        elifs_conditions = (
+            jax.numpy.array(elifs_conditions) if elifs_conditions else jax.numpy.empty(0)
+        )
```
Comment (dime10) [inline pennylane/ops/op_math/condition.py:562]: Why is this needed, just for convenience?

**Review by dime10 — CHANGES_REQUESTED** [review]
Nice work @PietropaoloFrisoni 💪

File: pennylane/ops/op_math/condition.py [hunk 5]
```diff
@@ -430,3 +446,128 @@ def wrapper(*args, **kwargs):
...
+
+        def run_jaxpr(jaxpr, *args):
+            return jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, *args)
+
+        def true_branch(args):
+            return run_jaxpr(jaxpr_true, *args)
+
+        def elif_branch(args, elifs_conditions, jaxpr_elifs):
+            if not jaxpr_elifs:
+                return false_branch(args)
```
[the remark this answers is not in the document]
Comment (PietropaoloFrisoni) [reply pennylane/ops/op_math/condition.py:472]: > since I wouldn't expect the recursion depth to be very large.

Exactly, I don't expect that the stack allocation ever becomes a problem for the number of `elifs` conditions that we usually provide to this function

File: pennylane/ops/op_math/condition.py (hunk 3 above)
Comment (PietropaoloFrisoni) [reply pennylane/ops/op_math/condition.py:548]: Good catch!

File: pennylane/ops/op_math/condition.py (hunk 4 above)
Comment (PietropaoloFrisoni) [reply pennylane/ops/op_math/condition.py:562]: `elifs` conditions are traced (as the first argument of the `qml.cond` function). Therefore, we pass them as a positional argument the the primitive, but if we try to pass them as a `list` or a `tuple` the execution fails because these are not standard JAX types.
This is why we have to turn such conditions into a JAX array. Am I missing something?

File: pennylane/ops/op_math/condition.py [hunk 6]
```diff
@@ -430,3 +446,128 @@ def wrapper(*args, **kwargs):
...
+                    f"Mismatch in output abstract values in {branch_type} branch"
+                    f"{'' if index is None else ' #' + str(index)} at position {i}: "
+                    f"{outval} vs {expected_outval}"
+                )
+
+        outvals_true = jaxpr_true.out_avals
+
+        if jaxpr_false is not None:
+            outvals_false = jaxpr_false.out_avals
+            validate_abstract_values(outvals_false, outvals_true, "false")
```
[the remark this answers is not in the document]
Comment (PietropaoloFrisoni) [reply pennylane/ops/op_math/condition.py:519]: > I think this omits the error check for when outvals_true has an element but there is no jaxpr_false defined, which should also be an error.

This corresponds to the following example

```
@qml.qnode(dev)
def circuit(pred, arg1, arg2):

    def true_fn(arg1, arg2):
        return qml.RX(arg2, wires=0)

    qml.cond(pred > 0, true_fn)(arg1, arg2)

    return qml.expval(qml.PauliZ(wires=0))
```

which works if `true_fn` is executed, but it fails if the condition is false. So I think you are right: we should prevent this case from happening. Good catch!

File: pennylane/ops/op_math/condition.py (hunk 2 above)
Comment (PietropaoloFrisoni) [reply pennylane/ops/op_math/condition.py:490]: It's not so strange to me because each `elif` condition is logically inserted between `true_fn` and `false_fn`, so if there are no `elifs` conditions we simply move to the `false_branch`. However, I agree that maybe it's better to move this logic outside of a function called `elif_branch`

File: pennylane/ops/op_math/condition.py (hunk 4 above)
Comment (dime10) [reply pennylane/ops/op_math/condition.py:571]: Ah got it. Well there are other ways to pass multiple positional variadic arguments, you just have to unpack them (`*elifs_conditions`) in the function call. The only downside is on the other end of the call (in the function definition), you would have to separate out the multiple variadic arguments from a single list of values again. You can see patterns like this in the Catalyst primitives:
```py
(condition, *elif_conditions_and_args, elif_jaxprs=()):
    elif_conditions = elif_conditions_and_args[:len(elif_jaxprs)]
    args = elif_conditions_and_args[len(elif_jaxprs):]
```

But the be honest I don't know what is better yet so I'm not requesting any changes :)

File: pennylane/ops/op_math/condition.py [hunk 7]
```diff
@@ -112,6 +114,15 @@ def cond(condition, true_fn, false_fn=None, elifs=()):
...
     branch determined at runtime. For more details, please see :func:`catalyst.cond`.
 
+    When used with :func:`~.pennylane.capture.enabled`, this function allows for general
+    if-elif-else constructs. As with the JIT mode, all branches will be captured,
+    with the executed branch determined at runtime.
+    However, the function cannot branch on mid-circuit measurements.
+    Each branch can receive arguments, but the arguments must be the same for all branches.
+    Both the arguments and the branches must be JAX-compatible.
+    If a branch returns one or more variables, every other branch must return the same abstract values.
+    If a branch returns one or more operators, these will be appended to the QueuingManager.
```
[the remark this answers is not in the document]
Comment (PietropaoloFrisoni) [reply pennylane/ops/op_math/condition.py]: At present, I am not touching the queue explicitly, as the branch that is supposed to be executed in the primitive definition is simply evaluated with `eval_jaxpr` and returned. This should be compatible with the default structure implemented so far for plxpr, but @albi3ro can correct me if I'm missing something

[the remark this answers is not in the document]
Comment (dime10) [reply pennylane/ops/op_math/condition.py]: The wording still seems strange to me. Looking at one of your examples:
```py
@qml.qnode(dev)
def circuit(pred, arg1, arg2):
    """Quantum circuit with conditional branches."""

    qml.RX(0.10, wires=0)

    def true_fn(arg1, arg2):
        qml.RY(arg1, wires=0)
        qml.RX(arg2, wires=0)
        qml.RZ(arg1, wires=0)

    def false_fn(arg1, arg2):
        qml.RX(arg1, wires=0)
        qml.RX(arg2, wires=0)

    def elif_fn1(arg1, arg2):
        qml.RZ(arg2, wires=0)
        qml.RX(arg1, wires=0)

    qml.cond(pred > 0, true_fn, false_fn, elifs=(pred == -1, elif_fn1))(arg1, arg2)
    qml.RX(0.10, wires=0)
    return qml.expval(qml.PauliZ(wires=0))
```

We can see that none of the branches return an operator, yet I'm assuming those gates in the branch functions still execute/queue correctly.

[the remark this answers is not in the document]
Comment (dime10) [reply pennylane/ops/op_math/condition.py]: Or are you saying this in a non-exclusive way, that is regardless of wether the operators are returned from a branch function or not, they will be executed?

[the remark this answers is not in the document]
Comment (PietropaoloFrisoni) [reply pennylane/ops/op_math/condition.py]: > regardless of whether the operators are returned from a branch function or not, they will be executed?

Yes, if that branch is supposed to be executed. It seems to me that this is compatible with the current usage of `qml.cond`. That is, operators are executed even if they are not returned:

```
@qml.qjit
@qml.qnode(dev)
def circuit(x):

    def true_fn():
        qml.RX(x, wires=0)

    qml.cond(x > 0, true_fn)()
    return qml.expval(qml.Z(0))
    
>>> circuit(1.34)
Array(0.22875281, dtype=float64)    
```

[the remark this answers is not in the document]
Comment (dime10) [reply pennylane/ops/op_math/condition.py]: Got it, what was confusing to me is that I read the sentence as "the operator is applied exclusively if it is returned".

[the remark this answers is not in the document]
Comment (PietropaoloFrisoni) [reply pennylane/ops/op_math/condition.py]: Done

File: pennylane/ops/op_math/condition.py [hunk 8]
```diff
@@ -100,7 +102,7 @@ def adjoint(self):
         return Conditional(self.meas_val, self.base.adjoint())
 
 
-def cond(condition, true_fn, false_fn=None, elifs=()):
+def cond(condition, true_fn: Callable, false_fn: Callable = None, elifs=()):
```
Comment (albi3ro) [inline pennylane/ops/op_math/condition.py:105]:
```suggestion
def cond(condition, true_fn: Callable, false_fn: Optional[Callable] = None, elifs=()):
```

File: pennylane/ops/op_math/condition.py [hunk 9]
```diff
@@ -112,6 +114,16 @@ def cond(condition, true_fn, false_fn=None, elifs=()):
     will be captured by Catalyst, the just-in-time (JIT) compiler, with the executed
     branch determined at runtime. For more details, please see :func:`catalyst.cond`.
 
+    When used with :func:`~.pennylane.capture.enabled`, this function allows for general
+    if-elif-else constructs. As with the JIT mode, all branches are captured,
+    with the executed branch determined at runtime.
+    However, the function cannot branch on mid-circuit measurements.
```
Comment (albi3ro) [inline pennylane/ops/op_math/condition.py:120]: Not sure this is worth writing out.  @mudit2812 's work on mcm's will make it so that measurement values are captured as booleans and compatible with this PR. So we will be able to integrate the two together.

File: pennylane/ops/op_math/condition.py [hunk 10]
```diff
@@ -129,13 +141,13 @@ def cond(condition, true_fn, false_fn=None, elifs=()):
     Args:
         condition (Union[.MeasurementValue, bool]): a conditional expression involving a mid-circuit
            measurement value (see :func:`.pennylane.measure`). This can only be of type ``bool`` when
-           decorated by :func:`~.qjit`.
+           decorated by :func:`~.qjit` or when using :func:`~.pennylane.capture.enabled`.
```
Comment (albi3ro) [inline pennylane/ops/op_math/condition.py:144]:
```suggestion
           decorated by :func:`~.qjit`.
```

Given:

1) capture is an experimental project I don't think we want to publicly expose too much at this point 
2) We will be able to capture `qml.cond(qml.measure(0), qml.RX)(x, 0)` once we can capture MCM's.

File: pennylane/ops/op_math/condition.py [hunk 11]
```diff
@@ -112,6 +114,16 @@ def cond(condition, true_fn, false_fn=None, elifs=()):
     will be captured by Catalyst, the just-in-time (JIT) compiler, with the executed
     branch determined at runtime. For more details, please see :func:`catalyst.cond`.
 
+    When used with :func:`~.pennylane.capture.enabled`, this function allows for general
```
Comment (albi3ro) [inline pennylane/ops/op_math/condition.py:117]: Given the capture project is experimental and internal at this point, I don't want to confuse and complicate this docstring for users.  Maybe we can have a little toggle at the bottom of the docstring for "Experimental program capture considerations" or something.  The part of the of the docstring above `Args:` should usually be just 1-2 sentences. 

 So yes, we have added a bunch of extra information here for qjit, but I don't think this is the place for program capture information.

File: pennylane/ops/op_math/condition.py [hunk 12]
```diff
@@ -430,3 +446,138 @@ def wrapper(*args, **kwargs):
...
+
+        return false_branch(args) if elif_branch_out is None else elif_branch_out
+
+    @cond_prim.def_abstract_eval
+    def _(*_, jaxpr_true, jaxpr_false, jaxpr_elifs):
+
+        # We check that the return values in each branch (true, and possibly false and elifs)
+        # have the same abstract values.
+        # The error messages are detailed to help debugging
+        def validate_abstract_values(
```
Comment (albi3ro) [inline pennylane/ops/op_math/condition.py:501]: Could this be extracted out of `_get_cond_qfunc_prim`? Looks like it doesn't depend on any closure variables.  

Purely stylistic and personal preference, but I try to reduce nested functions when possible.  Extracting them globally makes them much easier to debug, since they can be imported and run by themselves. It also makes things more readable as you don't have to worry about the namespace and closure variables.

File: pennylane/ops/op_math/condition.py [hunk 13]
```diff
@@ -430,3 +446,138 @@ def wrapper(*args, **kwargs):
...
+    @cond_prim.def_impl
+    def _(condition, elifs_conditions, *args, jaxpr_true, jaxpr_false, jaxpr_elifs):
+
+        def run_jaxpr(jaxpr, *args):
+            return jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, *args)
+
+        def true_branch(args):
+            return run_jaxpr(jaxpr_true, *args)
+
+        def elif_branch(args, elifs_conditions, jaxpr_elifs):
```
Comment (albi3ro) [inline pennylane/ops/op_math/condition.py:470]: I'm curious about the choice to use recursion instead of iteration.

Again, stylistic, non-blocking personal preference.

**Review by albi3ro — COMMENTED** [review]
One suggestion to try.

What would you think of combining `jaxpr_true`, `jaxpr_false`, and all the `jaxpr_elifs`? Same with combining the `condition` with the `elifs_conditions`.

We could essentially have:
```
jaxpr_branches = (jaxpr_true, *jaxpr_elifs, jaxpr_false)
conditions = jax.numpy.array([condition, *elifs_conditions, True])
```
And then the implementation could just be something like:
```
for condition, branch in zip(conditions, branches):
    if condition:
        return jax.core.eval_jaxpr(branch, consts, *args)
```

File: pennylane/ops/op_math/condition.py (hunk 11 above)
Comment (PietropaoloFrisoni) [reply pennylane/ops/op_math/condition.py]: Thanks for the suggestions. I agree with pretty much all the points. I left some comments so that it is easier to debug and is not a bunch of spaghetti.

As for this point, I moved the discussion to the last expandable section before `Args`, and it's just a couple of sentences.

**Comment by mudit2812** [comment]
@PietropaoloFrisoni if you think this is in a reasonably good state, I would like to use this as a base for my PR so that I can capture `qml.cond` with predicates that use mid-circuit measurements.

**Comment by PietropaoloFrisoni** [comment]
@mudit2812 Sure, no problem for me. I'll tell you if I need to make relevant changes, but hopefully, this is not the case

File: pennylane/ops/op_math/condition.py [hunk 14]
```diff
@@ -430,3 +446,106 @@ def wrapper(*args, **kwargs):
...
+    import jax  # pylint: disable=import-outside-toplevel
+
+    cond_prim = jax.core.Primitive("cond")
+    cond_prim.multiple_results = True
+
+    @cond_prim.def_impl
+    def _(conditions, *args, jaxpr_branches):
+
+        for pred, jaxpr in zip(conditions, jaxpr_branches):
+            if pred and jaxpr is not None:
```
Comment (dime10) [inline pennylane/ops/op_math/condition.py:485]: Is `jaxpr is None` allowed at this point?

Comment (PietropaoloFrisoni) [reply pennylane/ops/op_math/condition.py:485]: Yes if the false branch has not been provided

File: pennylane/ops/op_math/condition.py [hunk 15]
```diff
@@ -430,3 +446,127 @@ def wrapper(*args, **kwargs):
...
+
+
+def _capture_cond(condition, true_fn, false_fn=None, elifs=()) -> Callable:
+    """Capture compatible way to apply conditionals."""
+
+    import jax  # pylint: disable=import-outside-toplevel
+
+    cond_prim = _get_cond_qfunc_prim()
+
+    elifs = (elifs,) if len(elifs) > 0 and not isinstance(elifs[0], tuple) else elifs
```
Comment (dime10) [inline pennylane/ops/op_math/condition.py:534]: What kind of input does this cover?

**Review by dime10 — APPROVED** [review]
Nice work!

File: pennylane/ops/op_math/condition.py (hunk 15 above)
Comment (dime10) [reply pennylane/ops/op_math/condition.py:534]: Nice : )

File: doc/releases/changelog-dev.md [hunk 16]
```diff
@@ -4,6 +4,9 @@
 
 <h3>New features since last release</h3>
 
+* The `qml.cond` function can be captured into plxpr.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:7]: Potentially worth combining with the entry for `adjoint` and `ctrl`.

File: pennylane/ops/op_math/condition.py [hunk 17]
```diff
@@ -430,3 +446,127 @@ def wrapper(*args, **kwargs):
...
+
+        consts_per_branch = []
+        start = 0
+        for n in n_consts_per_branch:
+            consts_per_branch.append(consts_flat[start : start + n])
+            start += n
+
+        for pred, jaxpr, consts in zip(conditions, jaxpr_branches, consts_per_branch):
+            if pred and jaxpr is not None:
+                return jax.core.eval_jaxpr(jaxpr.jaxpr, consts, *args)
```
Comment (albi3ro) [inline pennylane/ops/op_math/condition.py:495]: If I'm reading this right, we should be able to combine these two loops?

File: pennylane/ops/op_math/condition.py [hunk 18]
```diff
@@ -430,3 +446,127 @@ def wrapper(*args, **kwargs):
...
+        jaxpr_elifs = []
+
+        for pred, elif_fn in elifs:
+            elifs_conditions.append(pred)
+            jaxpr_elifs.append(jax.make_jaxpr(functools.partial(elif_fn, **kwargs))(*args))
+
+        conditions = jax.numpy.array([condition, *elifs_conditions, True])
+
+        jaxpr_branches = [jaxpr_true, *jaxpr_elifs, jaxpr_false]
+        jaxpr_consts = [jaxpr.consts if jaxpr is not None else () for jaxpr in jaxpr_branches]
```
Comment (albi3ro) [inline pennylane/ops/op_math/condition.py:556]:
```suggestion
        jaxpr_consts = (jaxpr.consts if jaxpr is not None else () for jaxpr in jaxpr_branches)
```
since we will only be using this to iterate over a single time, we can use a generator.

File: tests/capture/test_capture_cond.py
Comment (albi3ro) [inline tests/capture/test_capture_cond.py:1]: Can we include some tests where we create the associated `jaxpr`, and then evaluate it using `jax.core.eval_jaxpr`?

File: tests/capture/test_capture_cond.py [hunk 19]
```diff
@@ -0,0 +1,466 @@
...
+        )(arg)
+        assert np.allclose(result, expected), f"Expected {expected}, but got {result}"
+
+    @pytest.mark.parametrize(
+        "selector, arg, expected",
+        [
+            (1, 10, 20),  # True condition
+            (-1, 10, 9),  # Elif condition 1
+            (-2, 10, 8),  # Elif condition 2
+            (-3, 10, ()),  # No condition met
```
Comment (albi3ro) [inline tests/capture/test_capture_cond.py:102]: I thought this case was forbidden?

**Comment by mudit2812** [comment]
I created a branch locally that uses the changes from this PR as well as from https://github.com/PennyLaneAI/pennylane/pull/6015. Seems like after creating a primitive for `qml.measure` which has a scalar boolean abstract eval, using mid-circuit measurements with `qml.cond` just works out of the box for capture. It cannot be executed for the same reason why any other classical processing on MCMs can't be executed, but I guess I don't need to do any further work other than adding tests 😄

File: pennylane/ops/op_math/condition.py (hunk 17 above)
Comment (PietropaoloFrisoni) [reply pennylane/ops/op_math/condition.py]: Yep! it's quite embarrassing that I missed this, but thanks for noticing

File: tests/capture/test_capture_cond.py
Comment (PietropaoloFrisoni) [reply tests/capture/test_capture_cond.py:1]: I added an additional check by calling `jax.core.eval_jaxpr` for each test circuit. It seems to me it's not a problem to add those to the same test since we are testing exactly the same circuits. Do you agree?

File: tests/capture/test_capture_cond.py (hunk 19 above)
Comment (PietropaoloFrisoni) [reply tests/capture/test_capture_cond.py:102]: As discussed, this does not trigger an error because such a check occurs in the abstract definition, which is not executed here. It does not seem a problem to me if we allow this case when the abstract definition is not executed. What do you think?

File: tests/capture/test_capture_cond.py
Comment (albi3ro) [reply tests/capture/test_capture_cond.py:1]: Thanks! Mind adding a similiar capture -> execute test in `test_cond_true_elifs_false`, `test_cond_true_false`, and `test_cond_with_jax_array`?

Comment (PietropaoloFrisoni) [reply tests/capture/test_capture_cond.py:1]: Done!

**Review by albi3ro — APPROVED** [review]
Excellent work 🎉
